File: clients/go/internal/cli/tendcmd.go
package cli

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"daisugi-verify/internal/config"
	"daisugi-verify/internal/distill"
	"daisugi-verify/internal/embed/potion"
	"daisugi-verify/internal/llm"
	"daisugi-verify/internal/pathways"
	"daisugi-verify/internal/pystr"
	"daisugi-verify/internal/tracejournal"
)

func (e *Env) lookup(k string) (string, bool) {
	v, ok := e.env[k]
	return v, ok
}

func (e *Env) potionEnv() potion.Env {
	return potion.Env{Lookup: e.lookup, Home: e.home, Notice: func(s string) { e.errf("%s\n", s) }}
}

func (e *Env) llmClient() *llm.Client {
	return llm.New(llm.Env{Getenv: e.lookup, Home: e.home, Stdout: e.Stdout, Stderr: e.Stderr,
		LookPath: func(name string) (string, error) { return lookPath(name, e.env["PATH"]) }})
}

// lookPath is shutil.which over a PATH value.
func lookPath(name, path string) (string, error) {
	for _, dir := range filepath.SplitList(path) {
		if dir == "" {
			dir = "."
		}
		p := filepath.Join(dir, name)
		if st, err := os.Stat(p); err == nil && !st.IsDir() && st.Mode()&0o111 != 0 {
			return p, nil
		}
	}
	return "", os.ErrNotExist
}

// matcher is the configured matcher, or the reason tend cannot run.
// refuse is set for a matcher this binary does not carry (PW-1);
// notBuilt for a key nothing builds, which Python raises on.
func (e *Env) matcher() (m *pathways.Matcher, notBuilt string, err error) {
	cfg, err := config.Load(filepath.Join(e.home, ".opendaisugi", "config.yaml"))
	if err != nil {
		return nil, "", fmt.Errorf("the config file is not one this binary reads: %w", err)
	}
	m, err = pathways.SelectMatcher(cfg.MatcherModel, e.potionEnv())
	if err != nil {
		return nil, "", err
	}
	if m == nil {
		return nil, cfg.MatcherModel, nil
	}
	return m, "", nil
}

func (e *Env) tend(args []string) error {
	const cmd = "tend"
	opts := []opt{
		{names: []string{"--data-dir"}, value: true, metavar: "PATH", help: "Daisugi data directory."},
		{names: []string{"--model"}, value: true, metavar: "TEXT", help: "Model used for template generalization + improvement."},
		{names: []string{"--min-traces"}, value: true, metavar: "INTEGER", help: "Minimum cluster size to distill."},
		{names: []string{"--lookback-days"}, value: true, metavar: "INTEGER", help: "How far back to scan the journal."},
		{names: []string{"--dry-run"}, help: "Run the pipeline but do not store pathways."},
	}
	p, err := parseArgs(args, opts, 0)
	if err != nil {
		return e.usage(cmd, err)
	}
	if p.help {
		return e.cmdHelp(cmd, "", "Run the distiller. Scans successful traces and produces compiled pathways.", opts)
	}
	minTraces, err := clickInt(p, "--min-traces", 3)
	if err != nil {
		return e.usage(cmd, err)
	}
	lookback, err := clickInt(p, "--lookback-days", 30)
	if err != nil {
		return e.usage(cmd, err)
	}
	model := p.str("--model", "anthropic/claude-sonnet-4-20250514")
	dataDir := p.str("--data-dir", filepath.Join(e.home, ".opendaisugi"))
	dry := p.flag("--dry-run")
	rep, err := e.runTend(dataDir, model, minTraces, lookback, dry)
	if err != nil {
		return err
	}
	e.out("tend complete: created=%d updated=%d skipped=%d in %.1fs\n", rep.Created, rep.Updated, rep.Skipped, rep.seconds)
	if len(rep.Pathways) > 0 {
		e.out("  %d pathway(s): %s\n", len(rep.Pathways), strings.Join(rep.Pathways, ", "))
	}
	for _, w := range rep.Warnings {
		e.out("  warning: %s\n", w)
	}
	return nil
}

// raisedError is an error Daisugi() raises before tend runs: the CLI
// prints its message and exits 1, and auto-tend does not catch it.
type raisedError struct{ msg string }

func (r *raisedError) Error() string { return r.msg }

type tendReport struct {
	distill.Report
	seconds float64
}

// tendState is what the read checks of a tend found, before it writes.
type tendState struct {
	m        *pathways.Matcher // nil for a matcher the binary does not carry
	identity string
	client   *llm.Client
	since    float64
	traces   []tracejournal.Distillable
}

// identityOf is active_model_name for a matcher key.
func identityOf(m *pathways.Matcher, key string) string {
	if m != nil {
		return m.Identity
	}
	if key == "int8" {
		return "all-MiniLM-L6-v2-int8"
	}
	return key
}

// tendCheck is every refusal of a tend run, made before anything is
// written: the matcher, the model, the journal's traces and refinements
// and the store's rows. extra is how many traces the caller will add
// first (auto-tend's conversions). A matcher the binary does not carry is
// refused only when the run would load its embedder: a stale row to
// re-embed, or enough traces to cluster.
func (e *Env) tendCheck(dataDir, model string, minTraces, lookback int64, dry bool, extra int, started float64) (*tendState, error) {
	const cmd = "tend"
	cfg, cerr := config.Load(filepath.Join(e.home, ".opendaisugi", "config.yaml"))
	if cerr != nil {
		return nil, e.refuse(cmd, fmt.Errorf("the config file is not one this binary reads: %w", cerr))
	}
	m, notCarried := pathways.SelectMatcher(cfg.MatcherModel, e.potionEnv())
	if notCarried == nil && m == nil {
		// Daisugi() resolves the matcher's threshold before it makes
		// anything, and raises for a key nothing builds; the CLI prints
		// the error's message and exits 1.
		return nil, &raisedError{"matcher_model=" + pystr.Repr(cfg.MatcherModel) + " is not a built embedder."}
	}
	st := &tendState{m: m, identity: identityOf(m, cfg.MatcherModel), client: e.llmClient(),
		since: started - float64(lookback)*86400}
	if err := st.client.Check(model); err != nil {
		return nil, e.refuse(cmd, err)
	}
	if fi, err := os.Stat(dataDir); err == nil && !fi.IsDir() {
		e.errf("daisugi %s: FileExistsError: %s exists and is not a directory\n", cmd, dataDir)
		return nil, exit(1)
	}
	if _, err := os.Stat(filepath.Join(dataDir, "journal", "index.db")); err == nil {
		// Read only: an old journal is migrated when tend runs, after
		// every refusal.
		j, err := tracejournal.OpenReadOnly(dataDir)
		if err != nil {
			return nil, e.failPy(cmd, err)
		}
		defer j.Close()
		if st.traces, err = j.ListSuccessful(&st.since); err != nil {
			return nil, e.tendErr(err)
		}
		if err := (&distill.Distiller{J: j}).Prepare(st.traces); err != nil {
			return nil, e.tendErr(err)
		}
	}
	stale := false
	if !dry {
		if _, err := os.Stat(filepath.Join(dataDir, "pathways.db")); err == nil {
			// Read only, as the journal: the store is migrated after
			// every refusal.
			s, err := pathways.OpenReadOnly(filepath.Join(dataDir, "pathways.db"))
			if err != nil {
				return nil, e.storeErr(cmd, err)
			}
			_, rerr := s.ReadAll(func(string) bool { return false })
			rows, aerr := s.All()
			s.Close()
			if errors.Is(rerr, pathways.ErrUnreadable) {
				return nil, e.refuse(cmd, rerr)
			}
			if aerr != nil {
				return nil, e.storeErr(cmd, aerr)
			}
			for _, r := range rows {
				if r["embedding_model"] != st.identity || r["embedding_model_version"] != pathways.EmbeddingModelVersion {
					stale = true
				}
			}
		}
	}
	if notCarried != nil && (stale || int64(len(st.traces)+extra) >= minTraces) {
		return nil, e.refuse(cmd, notCarried)
	}
	return st, nil
}

// runTend is Daisugi(...).tend(min_traces, lookback_days): the checks
// first, with nothing written; then the envelope cache, the journal and
// the store opened (and made), and the distiller run.
func (e *Env) runTend(dataDir, model string, minTraces, lookback int64, dry bool) (tendReport, error) {
	const cmd = "tend"
	started := nowSeconds()
	st, err := e.tendCheck(dataDir, model, minTraces, lookback, dry, 0, started)
	var raised *raisedError
	if errors.As(err, &raised) {
		e.errf("%s\n", raised.msg)
		e.raised = true
		return tendReport{}, exit(1)
	}
	if err != nil {
		return tendReport{}, err
	}
	// From here on the run writes, so input this binary cannot read stops
	// it without the refusal's "Nothing was changed.".
	if err := tracejournal.EnsureEnvelopeCache(filepath.Join(dataDir, "envelope_cache.db")); err != nil {
		return tendReport{}, e.afterWrites(cmd, err)
	}
	j, err := tracejournal.Open(dataDir)
	if err != nil {
		return tendReport{}, e.afterWrites(cmd, err)
	}
	defer j.Close()
	d := &distill.Distiller{J: j, M: st.m, LLM: st.client,
		Opt: distill.Options{Model: model, MinTraces: minTraces, LookbackDays: lookback, ValidationSplit: 0.6,
			StructureWeight: 0.5, Now: nowSeconds, Z3TimeoutMs: 500, Identity: st.identity}}
	if st.m != nil {
		d.Opt.Threshold = st.m.Threshold
	}
	if err := d.Prepare(st.traces); err != nil {
		return tendReport{}, e.afterWrites(cmd, err)
	}
	if dry {
		d.S = &distill.MemStore{}
	} else {
		s, err := pathways.Open(filepath.Join(dataDir, "pathways.db"))
		if err != nil {
			return tendReport{}, e.afterWrites(cmd, err)
		}
		defer s.Close()
		d.S = distill.DiskStore{Store: s}
	}
	rep, err := d.Tend(st.traces)
	if err != nil {
		return tendReport{}, e.afterWrites(cmd, err)
	}
	return tendReport{Report: rep, seconds: nowSeconds() - started}, nil
}

func (e *Env) tendErr(err error) error {
	if errors.Is(err, tracejournal.ErrUnreadable) || errors.Is(err, pathways.ErrUnreadable) {
		return e.refuse("tend", err)
	}
	return e.failPy("tend", err)
}

// afterWrites is tendErr, except that once a run has started writing,
// input this binary cannot read stops it (exit 2) with no claim that
// nothing changed. The checks before the first write read the same
// input, so no case reaches this.
func (e *Env) afterWrites(cmd string, err error) error {
	if errors.Is(err, tracejournal.ErrUnreadable) || errors.Is(err, pathways.ErrUnreadable) {
		e.errf("daisugi %s: %v. The run stopped after its first write.\n", cmd, err)
		return exit(2)
	}
	return e.failPy(cmd, err)
}

func (e *Env) distillRepeats(args []string) error {
	const cmd = "distill-repeats"
	opts := []opt{
		{names: []string{"--data-dir"}, value: true, metavar: "PATH", help: "Daisugi data directory (reads <data-dir>/gateway/turns.jsonl)."},
		{names: []string{"--top"}, value: true, metavar: "INTEGER", help: "Maximum number of rows to print."},
	}
	p, err := parseArgs(args, opts, 0)
	if err != nil {
		return e.usage(cmd, err)
	}
	if p.help {
		return e.cmdHelp(cmd, "", "Rank repeated gateway asks into a reuse worklist.", opts)
	}
	top, err := clickInt(p, "--top", 20)
	if err != nil {
		return e.usage(cmd, err)
	}
	dataDir := p.str("--data-dir", filepath.Join(e.home, ".opendaisugi"))
	path := filepath.Join(dataDir, "gateway", "turns.jsonl")
	// Skipped lines are logged on the opendaisugi logger, which prints
	// nothing by default.
	turns, _, err := distill.LoadTurns(path)
	if err != nil {
		return e.tendErrAs(cmd, err)
	}
	if len(turns) == 0 {
		e.out("no turns recorded yet — run `daisugi gateway` to start journaling.\n")
		return nil
	}
	signed := false
	for _, t := range turns {
		if t.Signature != "" {
			signed = true
		}
	}
	var cands []distill.Candidate
	if signed {
		m, notBuilt, err := e.matcher()
		if err != nil {
			return e.refuse(cmd, err)
		}
		if notBuilt != "" {
			e.errf("matcher_model=%s is not a built embedder.\n", pystr.Repr(notBuilt))
			return exit(1)
		}
		emb, err := m.Embedder()
		if err != nil {
			e.errf("%v\n", err)
			return exit(2)
		}
		var find func(string) bool
		db := filepath.Join(dataDir, "pathways.db")
		if _, err := os.Stat(db); err == nil {
			s, err := pathways.Open(db)
			if err != nil {
				return e.storeErr(cmd, err)
			}
			defer s.Close()
			key := m.Key
			find = func(task string) bool {
				r, err := s.Find(task, key, e.potionEnv(), -1)
				return err == nil && r.Match != nil
			}
		}
		cands = distill.RankReuse(turns, emb, m.Threshold, find)
	}
	if len(cands) == 0 {
		e.out("no repeated asks found yet.\n")
		return nil
	}
	e.out("%4s  %4s  %10s  %9s  reusable?  task\n", "rank", "occ", "tokens", "dollars")
	// candidates[:top], a negative top counting from the end.
	n := int64(len(cands))
	if top < 0 {
		n = max(0, n+top)
	} else if top < n {
		n = top
	}
	for i, c := range cands[:n] {
		e.out("%s\n", c.Row(i+1))
	}
	return nil
}

func (e *Env) tendErrAs(cmd string, err error) error {
	if errors.Is(err, tracejournal.ErrUnreadable) || errors.Is(err, pathways.ErrUnreadable) {
		return e.refuse(cmd, err)
	}
	return e.failPy(cmd, err)
}

File: clients/rust/src/cli/tendcmd.rs
//! `daisugi tend` and `daisugi distill-repeats`: the distiller and the
//! gateway worklist, with the flags, output, exit codes and files of the
//! Python CLI. Every refusal comes before the first write.

use std::cell::RefCell;
use std::rc::Rc;

use super::gardencmd::now_seconds;
use super::{exit, parse_args, Env, Opt, Res, Stop};
use crate::distill::repeats::{load_turns, rank_reuse};
use crate::distill::{load_embedder, prepare, Distiller, Options, Report, Sink, TendErr};
use crate::gate::py::text::repr;
use crate::llm::Client;
use crate::pathways::find::{select_matcher, Matcher, EMBEDDING_MODEL_VERSION};
use crate::pathways::potion;
use crate::pathways::store::{Col, Store};
use crate::pathways::PwErr;
use crate::tracejournal::{ensure_envelope_cache, Distillable, Journal};

/// The model tend asks by default.
pub const DEFAULT_MODEL: &str = "anthropic/claude-sonnet-4-20250514";

/// How auto-tend reports a tend that raised: "tend failed: <exception
/// class name>: <message>".
pub fn tend_failure(msg: &str) -> String {
    if let Some((head, rest)) = msg.split_once(": ") {
        if !head.contains(' ') {
            let name = head.rsplit('.').next().unwrap_or(head);
            return format!("tend failed: {name}: {rest}");
        }
    }
    format!("tend failed: {msg}")
}

/// What the read checks of a tend found, before it writes.
pub struct TendState {
    matcher: Option<Matcher>,
    identity: String,
    client: Client,
    traces: Vec<Distillable>,
}

/// A tend report and how long the run took.
pub struct TendReport {
    pub rep: Report,
    pub seconds: f64,
}

/// Why the checks of a tend stopped it.
pub enum CheckStop {
    /// `Daisugi()` raised before tend ran: the CLI prints the message and
    /// exits 1, and auto-tend does not catch it.
    Raised(String),
    Stop(Stop),
}

impl From<Stop> for CheckStop {
    fn from(s: Stop) -> Self {
        CheckStop::Stop(s)
    }
}

/// `active_model_name` for a matcher key.
fn identity_of(m: &Option<Matcher>, key: &str) -> String {
    match m {
        Some(m) => m.identity().to_string(),
        None if key == "int8" => "all-MiniLM-L6-v2-int8".into(),
        None => key.to_string(),
    }
}

impl Env {
    /// The potion settings of this run; a download notice goes to `notes`.
    pub(super) fn potion_env(&self, notes: &Rc<RefCell<String>>) -> potion::Env {
        let n = notes.clone();
        let mut pe = potion::Env::from_process(self.env.clone(), self.home.clone());
        pe.notice = Box::new(move |s| {
            n.borrow_mut().push_str(s);
            n.borrow_mut().push('\n');
        });
        pe
    }

    fn drain_notes(&mut self, notes: &Rc<RefCell<String>>) {
        let text = std::mem::take(&mut *notes.borrow_mut());
        self.errf(&text);
    }

    /// Prints what the model client printed, then empties it.
    pub(super) fn drain_client(&mut self, c: &mut Client) {
        let out = std::mem::take(&mut c.stdout);
        let err = std::mem::take(&mut c.stderr);
        self.out(&out);
        self.errf(&err);
    }

    pub(super) fn tend(&mut self, args: &[String]) -> Res {
        const CMD: &str = "tend";
        let opts = [
            Opt::val(&["--data-dir"], "PATH", "Daisugi data directory."),
            Opt::val(&["--model"], "TEXT", "Model used for template generalization + improvement."),
            Opt::val(&["--min-traces"], "INTEGER", "Minimum cluster size to distill."),
            Opt::val(&["--lookback-days"], "INTEGER", "How far back to scan the journal."),
            Opt::flag(&["--dry-run"], "Run the pipeline but do not store pathways."),
        ];
        let p = parse_args(args, &opts, 0).map_err(|m| self.usage_stop(CMD, m))?;
        if p.help {
            return self.cmd_help(CMD, "", "Run the distiller. Scans successful traces and produces compiled pathways.", &opts);
        }
        let min_traces = self.click_int(CMD, &p, "--min-traces", 3)?;
        let lookback = self.click_int(CMD, &p, "--lookback-days", 30)?;
        let model = p.str("--model", DEFAULT_MODEL);
        let data_dir = p.str("--data-dir", &format!("{}/.opendaisugi", self.home));
        let r = self.run_tend(&data_dir, &model, min_traces, lookback, p.flag("--dry-run"))?;
        let mut b = format!(
            "tend complete: created={} updated={} skipped={} in {:.1}s\n",
            r.rep.created, r.rep.updated, r.rep.skipped, r.seconds
        );
        if !r.rep.pathways.is_empty() {
            b.push_str(&format!("  {} pathway(s): {}\n", r.rep.pathways.len(), r.rep.pathways.join(", ")));
        }
        for w in &r.rep.warnings {
            b.push_str(&format!("  warning: {w}\n"));
        }
        self.out(&b);
        Ok(())
    }

    /// Every refusal of a tend run, made before anything is written: the
    /// matcher, the model, the journal's traces and refinements and the
    /// store's rows. `extra` is how many traces the caller adds first
    /// (auto-tend's conversions). A matcher the binary does not carry is
    /// refused only when the run would load its embedder: a stale row to
    /// re-embed, or enough traces to cluster.
    #[allow(clippy::too_many_arguments)]
    pub(super) fn tend_check(
        &mut self,
        data_dir: &str,
        model: &str,
        min_traces: i128,
        lookback: i128,
        dry: bool,
        extra: usize,
        started: f64,
        notes: &Rc<RefCell<String>>,
    ) -> Result<TendState, CheckStop> {
        const CMD: &str = "tend";
        let cfg = match super::config::load(&format!("{}/.opendaisugi/config.yaml", self.home)) {
            Ok(c) => c,
            Err(e) => return Err(self.refuse(CMD, &format!("the config file is not one this binary reads: {e}")).unwrap_err().into()),
        };
        let pe = self.potion_env(notes);
        let (matcher, not_carried) = match select_matcher(&cfg.matcher_model, &pe) {
            Ok(Some(m)) => (Some(m), None),
            // Daisugi() resolves the matcher's threshold before it makes
            // anything, and raises for a key nothing builds.
            Ok(None) => return Err(CheckStop::Raised(format!("matcher_model={} is not a built embedder.", repr(&cfg.matcher_model)))),
            Err(nc) => (None, Some(nc.0)),
        };
        let identity = identity_of(&matcher, &cfg.matcher_model);
        let client = Client::new(self.env.clone(), &self.home);
        if let Err(why) = client.check(model) {
            return Err(self.refuse(CMD, &why).unwrap_err().into());
        }
        if std::fs::metadata(data_dir).map(|m| !m.is_dir()).unwrap_or(false) {
            self.errf(&format!("daisugi {CMD}: FileExistsError: {data_dir} exists and is not a directory\n"));
            return Err(Stop::Exit(1).into());
        }
        let since = started - lookback as f64 * 86_400.0;
        let mut traces = vec![];
        if std::fs::metadata(format!("{data_dir}/journal/index.db")).is_ok() {
            // Read only: an old journal is migrated when tend runs, after
            // every refusal.
            let j = Journal::open_read_only(data_dir).map_err(|e| self.pw_err(CMD, e))?;
            traces = j.list_successful(Some(since)).map_err(|e| self.pw_err(CMD, e))?;
            prepare(&j, &traces).map_err(|e| self.pw_err(CMD, e))?;
        }
        let mut stale = false;
        let db = format!("{data_dir}/pathways.db");
        if !dry && std::fs::metadata(&db).is_ok() {
            // Read only, as the journal: the store is migrated after every
            // refusal.
            let s = Store::open_read_only(&db).map_err(|e| self.pw_err(CMD, e))?;
            if let Err(e @ PwErr::Unreadable(_)) = s.read_all(&|_| false) {
                return Err(self.pw_err(CMD, e).into());
            }
            let rows = s.provenance().map_err(|e| self.pw_err(CMD, e))?;
            let is = |c: &Col, v: &str| matches!(c, Col::Text(t) if t == v);
            stale = rows.iter().any(|r| !is(&r[2], &identity) || !is(&r[3], EMBEDDING_MODEL_VERSION));
        }
        if let Some(why) = not_carried {
            if stale || (traces.len() + extra) as i128 >= min_traces {
                return Err(self.refuse(CMD, &why).unwrap_err().into());
            }
        }
        Ok(TendState { matcher, identity, client, traces })
    }

    /// `Daisugi(...).tend(min_traces, lookback_days)`: the checks first,
    /// with nothing written; then the envelope cache, the journal and the
    /// store opened (and made), and the distiller run.
    pub(super) fn run_tend(&mut self, data_dir: &str, model: &str, min_traces: i128, lookback: i128, dry: bool) -> Result<TendReport, Stop> {
        const CMD: &str = "tend";
        let started = now_seconds();
        let notes = Rc::new(RefCell::new(String::new()));
        let st = match self.tend_check(data_dir, model, min_traces, lookback, dry, 0, started, &notes) {
            Ok(st) => st,
            Err(CheckStop::Raised(msg)) => {
                self.errf(&format!("{msg}\n"));
                self.raised = true;
                return Err(Stop::Exit(1));
            }
            Err(CheckStop::Stop(s)) => return Err(s),
        };
        let TendState { matcher, identity, mut client, traces } = st;
        // From here on the run writes, so input this binary cannot read
        // stops it without the refusal's "Nothing was changed.".
        ensure_envelope_cache(&format!("{data_dir}/envelope_cache.db")).map_err(|e| self.after_writes(CMD, e))?;
        let j = Journal::open(data_dir).map_err(|e| self.after_writes(CMD, e))?;
        let prepared = prepare(&j, &traces).map_err(|e| self.after_writes(CMD, e))?;
        let store = if dry {
            None
        } else {
            Some(Store::open(&format!("{data_dir}/pathways.db")).map_err(|e| self.after_writes(CMD, e))?)
        };
        let pe = self.potion_env(&notes);
        let threshold = matcher.as_ref().map(|m| m.threshold()).unwrap_or(0.0);
        let rep = {
            let mut d = Distiller {
                prepared,
                sink: match &store {
                    Some(s) => Sink::Disk(s),
                    None => Sink::Mem(vec![]),
                },
                matcher,
                pe: &pe,
                emb: Default::default(),
                llm: &mut client,
                opt: Options {
                    model: model.to_string(),
                    min_traces,
                    lookback_days: lookback,
                    threshold,
                    validation_split: 0.6,
                    structure_weight: 0.5,
                    identity,
                    z3_timeout_ms: 500,
                },
                now: &now_seconds,
            };
            d.tend(&traces)
        };
        self.drain_notes(&notes);
        self.drain_client(&mut client);
        match rep {
            Ok(rep) => Ok(TendReport { rep, seconds: now_seconds() - started }),
            Err(TendErr::Pw(e)) => Err(self.after_writes(CMD, e)),
            Err(TendErr::Os(e)) => Err(self.pw_err(CMD, PwErr::Io(e))),
        }
    }

    /// An error once a run has written: as `pw_err`, except that input
    /// this binary cannot read stops the run (exit 2) with no claim that
    /// nothing changed. The checks before the first write read the same
    /// input, so no case reaches this.
    fn after_writes(&mut self, cmd: &str, e: PwErr) -> Stop {
        match e {
            PwErr::Unreadable(why) => {
                self.errf(&format!("daisugi {cmd}: {why}. The run stopped after its first write.\n"));
                Stop::Exit(2)
            }
            other => self.pw_err(cmd, other),
        }
    }

    pub(super) fn distill_repeats(&mut self, args: &[String]) -> Res {
        const CMD: &str = "distill-repeats";
        let opts = [
            Opt::val(&["--data-dir"], "PATH", "Daisugi data directory (reads <data-dir>/gateway/turns.jsonl)."),
            Opt::val(&["--top"], "INTEGER", "Maximum number of rows to print."),
        ];
        let p = parse_args(args, &opts, 0).map_err(|m| self.usage_stop(CMD, m))?;
        if p.help {
            return self.cmd_help(CMD, "", "Rank repeated gateway asks into a reuse worklist.", &opts);
        }
        let top = self.click_int(CMD, &p, "--top", 20)?;
        let data_dir = p.str("--data-dir", &format!("{}/.opendaisugi", self.home));
        // Skipped lines are logged on the opendaisugi logger, which prints
        // nothing by default.
        let turns = load_turns(&format!("{data_dir}/gateway/turns.jsonl")).map_err(|e| self.pw_err(CMD, e))?;
        if turns.is_empty() {
            self.out("no turns recorded yet \u{2014} run `daisugi gateway` to start journaling.\n");
            return Ok(());
        }
        let mut cands = vec![];
        if turns.iter().any(|t| !t.signature.is_empty()) {
            let cfg = match super::config::load(&format!("{}/.opendaisugi/config.yaml", self.home)) {
                Ok(c) => c,
                Err(e) => return self.refuse(CMD, &format!("the config file is not one this binary reads: {e}")),
            };
            let notes = Rc::new(RefCell::new(String::new()));
            let pe = self.potion_env(&notes);
            let m = match select_matcher(&cfg.matcher_model, &pe) {
                Ok(Some(m)) => m,
                Ok(None) => {
                    self.errf(&format!("matcher_model={} is not a built embedder.\n", repr(&cfg.matcher_model)));
                    return exit(1);
                }
                Err(nc) => return self.refuse(CMD, &nc.0),
            };
            let emb = load_embedder(&m, &pe);
            self.drain_notes(&notes);
            let emb = match emb {
                Ok(e) => e,
                Err(why) => {
                    self.errf(&format!("{why}\n"));
                    return exit(2);
                }
            };
            let db = format!("{data_dir}/pathways.db");
            let store = if std::fs::metadata(&db).is_ok() { Some(Store::open(&db).map_err(|e| self.pw_err(CMD, e))?) } else { None };
            let key = cfg.matcher_model.clone();
            let mut cache = crate::pathways::find::Cache::default();
            let mut find = |task: &str| -> bool {
                match &store {
                    Some(s) => matches!(s.find(task, &key, &pe, None, &mut cache), Ok(r) if r.matched.is_some()),
                    None => false,
                }
            };
            cands = rank_reuse(&turns, &emb, m.threshold(), &mut find);
            self.drain_notes(&notes);
        }
        if cands.is_empty() {
            self.out("no repeated asks found yet.\n");
            return Ok(());
        }
        let mut b = format!("{:>4}  {:>4}  {:>10}  {:>9}  reusable?  task\n", "rank", "occ", "tokens", "dollars");
        // candidates[:top], a negative top counting from the end.
        let n = cands.len() as i128;
        let n = if top < 0 { (n + top).max(0) } else { top.min(n) } as usize;
        for (i, c) in cands[..n].iter().enumerate() {
            b.push_str(&format!("{}\n", c.row(i + 1)));
        }
        self.out(&b);
        Ok(())
    }
}
